File: 面试高频题/0200 岛屿数量.py
from typing import List
import collections

class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        def bfs(grid,i,j):
            q = collections.deque()
            q.append((i,j))
            while q:
                i, j = q.popleft()
                if 0<=i<len(grid) and 0<=j<len(grid[0]) and grid[i][j]=='1':
                    grid[i][j] = '0'
                    q.extend([(i-1,j), (i+1,j), (i,j-1), (i,j+1)])
            cnt = 0
            for i in range(len(grid)):
                for j in range(len(grid[0])):
                    if grid[i][j] == '0':
                        continue
                    bfs(grid, i, j)
                    cnt += 1
            return cnt

File: 面试高频题/0200 岛屿数量.cpp
#include <iostream>
#include <vector>
#include <queue>

using namespace std;

class Solution {
public:
    vector<vector<char>> g;
    vector<vector<int>> d = {{0,1},{0,-1},{-1,0},{1,0}};
    void bfs(int i, int j) {
        queue<vector<int>> q;
        q.push({i,j});
        while (!q.empty()) {
            vector<int> tmp = q.front();
            q.pop();
            i=tmp[0], j=tmp[1];
            if (i>=0 && i<g.size() && j>=0 && j<g[0].size() && g[i][j] == '1') {
                g[i][j] = '0';
                for (auto const &x: d)
                    q.push({i+x[0], j+x[1]});
            }
            // g[i][j]='0';
            // for (auto const &x: d) {
            //     int ni=i+x[0], nj=j+x[1];
            //     if (ni>=0 && ni<g.size() && nj>=0 && nj<g[0].size() && g[ni][nj] == '1')
            //         q.push({ni,nj});
            // }
        }
    }

    int numIslands(vector<vector<char>>& grid) {
        if (grid.empty())
            return 0;
        int res = 0;
        g = grid;
        int rowNum = g.size();
        int colNum = g[0].size();
        for (int i=0; i<rowNum; ++i) {
            for (int j=0; j<colNum; ++j) {
                if (g[i][j] == '0')
                    continue;
                bfs(i, j);
                ++res;
            }
        }
        return res;
    }
};

int main() {
    vector<vector<char>> grid = {
        {'0','1','0','1','0'},
        {'1','1','0','1','0'},
        {'1','1','0','0','1'},
        {'0','0','0','1','1'}
    };

    grid = {
        {'1','1','0','0','0'},
        {'1','1','0','0','0'},
        {'0','0','1','0','0'},
        {'0','0','0','1','1'}
    };

    Solution sol;
    cout << sol.numIslands(grid) << endl;
    
    return 0;
}